src/merkle/merkle_tree.rs

```diff
@@ -19,31 +20,65 @@ impl MerkleTree {
     ///
     /// # Errors
     /// Returns an error if the number of leaves is smaller than two or is not a power of two.
-    pub fn new(leaves: Vec<Word>) -> Result<Self, MerkleError> {
+    pub fn new<T, I, V>(leaves: T) -> Result<Self, MerkleError>
+    where
+        T: IntoIterator<IntoIter = I>,
+        I: ExactSizeIterator + Iterator<Item = V>,
+        V: Borrow<Word>,
+    {
+        let leaves = leaves.into_iter();
         let n = leaves.len();
         if n <= 1 {
             return Err(MerkleError::DepthTooSmall(n as u8));
         } else if !n.is_power_of_two() {
             return Err(MerkleError::NumLeavesNotPowerOfTwo(n));
         }
 
-        // create un-initialized vector to hold all tree nodes
-        let mut nodes = unsafe { uninit_vector(2 * n) };
-        nodes[0] = [Felt::ZERO; 4];
+        // The tree contains `n` leaves, which is a power of two. The depth of the tree is `d`
+        // defined as `2^d=n` or `d=lb(n)`. The total number of elements in the tree is
+        // `2**(d+1)-1` or `2n-1` (since `n=2**d`).
+        //
+        // Below we compute the tree size `+1`, the first element is set to `0`, so that the length
+        // of `nodes` will be a power of two.
+        let final_capacity = 2 * n;
+
+        // Allocate data to accomodate all the elements, and get a reference to the underlying
+        // buffer
+        let mut nodes: Vec<Word> = Vec::with_capacity(final_capacity);
+        let buffer = nodes.spare_capacity_mut();
+
+        // The bottom layer goes to the higher indeces (at the end of the vector)
+        let last_layer = &mut buffer[n..];
+        for (pos, el) in leaves.enumerate() {
+            last_layer[pos].write(*el.borrow());
+        }
 
-        // copy leaves into the second part of the nodes vector
-        nodes[n..].copy_from_slice(&leaves);
+        let mut pos = n - 1;
+        let mut parent = final_capacity - 2;
+        while pos > 0 {
+            let left = unsafe { buffer[parent].assume_init() };
+            let right = unsafe { buffer[parent + 1].assume_init() };
```

Comment:
I'm not sure if the best outcome is to still have unsafe inside this function.

I thought we agreed on deciding what is the best strategy for `uninit_vec` so here would be just another consumer of this strategy - instead of implementing an isolated strategy of its own.

As discussed, unsafe can be a good thing sometimes, but there is no justification to use it here besides the inconsistency between `Word` and `RpoDigest`. I believe fixing the problem would be the same as going for the root cause instead of amending one of the branches.

Reply:
Humm, I had the goal of keep this [focused](https://github.com/0xPolygonMiden/crypto/pull/58/files#r1108445324). But lets see, these are [your proposals](https://github.com/0xPolygonMiden/crypto/pull/58#pullrequestreview-1301210043):

1. `fn<T>(usize) -> Vec<MaybeUninit<T>>`

I see no issue, I can open the PR if that is what we decide to go with, but note that this does not remove the `unsafe` block. My preference is still to use `spare_capacity_mut`, since the `uninit_vector` will boil down to a 2-liner and `spare_capacity_mut` is "easy" enough to use, but that is just a preference.

2. `unsafe fn<T>(usize) -> Vec<T>`

I feel like going over this will make us go in circles again. But, my interpretation of the Rust spec is that this is UB even for integer types, and for LLVM I believe these are poison values which are an inch away from UB. The issue with this is that 1. it doesn't remove the unsafe code anyways 2. it seems to be a dispute in the rust core language what to do with uninitialized data so there is not much we can rely on for this.

3. `fn(usize) -> Vec<Word>`

IMO this has the same problems as `2`, and it is a broken API since it is not a safe to use.

4. `fn<T>(usize) -> MaybeUninit<Vec<T>`

This is incorrect, this is saying the fat pointer `Vec` is uninitialized. This is not what we have, we get a valid vector from `Vec::with_capacity`. What we want to say is that the data contained in the vector storage is uninitialized, so it has to be `Vec<MaybeUninit<T>>`

So to summarize the above. IMO the only reasonable solution without guess work is `fn<T>(usize) -> Vec<MaybeUninit<T>>`, and IMO that is as good as `spare_capacity_mut`.

I really don't have anything else to add to this discussion.

Reply:
We can discard `2` and `4` for the reasons you pointed. I don't think `3` is invalid tho. We will have a safe function that will output a valid vector.

My vote would be to go for `3` so we don't need to delegate unnecessary `unsafe` code, but I'm ok if we go with `1`.

Reply:
> I don't think 3 is invalid tho

I really don't know what else to say. I have provided references to the language and my interpretation of why that is UB. Do you have anything to present to backup this opinion?

> It's worthy noting that for 3, how we implement that, is an implementation detail of the function. We could go completely safe and

both of these options zero the memory out, I'm fine with either, but note that the whole point of this discussion was to prevent zeroing the memory.

Reply:
I think I found an answer, and my rationale was incorrect.

This is undefined behavior:

```rust
let x: i32 = unsafe { MaybeUninit::uninit().assume_init() };
```

but this is not:

```rust
let mut v: Vec<i32> = Vec::with_capacity(1);
unsafe { v.set_len(1) }
```

The snippet above has an issue of safety, but not validity. If the code above was followed by a read of `v[0]`, then we would have UB caused because of the lack of safety. But as long as we don't read the memory, and properly initialize the memory `v[0] = 1i32` there wont be UB.

The reason is subtle and depends on the operational semantics of the language, which I don't fully understand it, but the language allows for `&mut uninit` and it was considered that forbidden it would cause UB on too much code that already exists.

So as long as we constraint the type to be `Copy` (which forbids a Drop), and the caller of the function initializes/writes to the complete buffer, we should be okay:

```rust
unsafe fn uninit_vector<T: Copy>(size: usize) -> Vec<T> {
    let mut v = Vec::with_capacity(size);
    unsafe { v.set_len(size) };
    v
}
```

References:

- https://rust-lang.github.io/rfcs/2930-read-buf.html
- also the issue 346 on rust-lang/unsafe-code-guidelines

Reply:
With the above said. If we are going to keep the `uninit_vector` I'm still of the opinion that we should keep the `unsafe` marker, specially since this function is exposed on a different crate.